scripts/controllers/RobotArm.py

```diff
@@ -0,0 +1,198 @@
+        # need a sleep, so message is sent from publisher, interestingly.
+        rospy.sleep(2)
+
+    def joint_names(self):
+        """
+        get the joint names of the robot
+        """
+        if self.data_exists:
+            return self.names
+
+    def joint_angles(self):
+        """
+        gets the joint angles of the robot arm
+        """
+        if self.data_exists:
+            return np.array(self.joint_data.position)[self.valid_indices]
+
+    def joint_velocities(self):
+        """
+        gets the joint velocities of the robot arm
+        """
+        if self.data_exists:
+            return np.array(self.joint_data.velocity)[self.valid_indices]
+
+    def joint_angle(self, joint_name):
+        """
+        gets the joint angle based on the provided joint name.
+        """
+        if self.data_exists:
+            angles = self.joint_angles()
+            return angles[self.mapping[joint_name]]
+
+    def joint_velocity(self, joint_name):
+        """
+        gets the joint velocity based on the provided joint name.
+        """
+        if self.data_exists:
+            velocities = self.joint_velocities()
+            return velocities[self.mapping[joint_name]]
+
+    def set_joint_position_speed(self, speed=1.0):
+        rospy.logerr('Set Joint Position Speed Not Implemented for Robot Arm')
+
+    def move_to_joint_positions(self, positions, sleep=1.0, timeout=15.0):
+        """
+        moves robot to specific joint positions.
+        """
+        if not rospy.is_shutdown():
+            self.controller_manager.switch_mode(ControllerType.POSITION)
+            if len(positions) != self.num_joints:
+                raise ValueError("Wrong number of joints provided.")
+            for index, pos in enumerate(positions):
+                self.pos_pubs[index].publish(Float64(pos))
+            rospy.sleep(sleep)
+
+    def set_joint_velocities(self, velocities):
+        """
+        affects the set joint velocities
+        """
+        self.controller_manager.switch_mode(ControllerType.VELOCITY)
+
+        if len(velocities) != self.num_joints:
+            raise ValueError("Wrong number of joints provided.")
+
+        for index, vel in enumerate(velocities):
+            self.vel_pubs[index].publish(Float64(vel))
+
+    def set_joint_trajectory(self, trajectories, time_per_step=1.0):
+        """
+        sends a joint trajectory command.
+        """
+        if not rospy.is_shutdown():
+            self.controller_manager.switch_mode(ControllerType.TRAJECTORY)
+
+            # 3d tensor of shape (3, n_joint, n_point)
+            points_tensor = np.array(trajectories)
+            assert points_tensor.shape[1] == 3, "3 row should be provided for pos, vel and acc."
+            assert len(points_tensor.shape) == 3, "Trajectories should be of 3-d shape!"
+            assert points_tensor.shape[2] == self.num_joints, "Trajectory should " \
```

Comment:
These as well, but is there any good coding practice for this?

Reply:
updated.